### ISSUE TITLE
** (ArgumentError) Postgrex expected %Postgrex.INET{}, got %Postgrex.CIDR

### ISSUE THREAD

**Issue by darkbushido** [elixir-ecto/postgrex#292]
```elixir
Repo.get_by(IpAddress, %{subnet: %Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}})
```
gives the following error
```elixir
** (ArgumentError) Postgrex expected %Postgrex.INET{}, got %Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}. Please make sure the value you are passing matches the definition in your table or in your query or convert the value accordingly.
[debug] QUERY ERROR source="ip_addresses" db=3.9ms
SELECT i0."id", i0."ip", i0."mac", i0."subnet", i0."inserted_at", i0."updated_at" FROM "ip_addresses" AS i0 WHERE (i0."subnet" = $1) [%Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}]
             (ecto) /Users/lance/rapid7/Xanatos-poc/xanatos/deps/postgrex/lib/postgrex/type_module.ex:716: Ecto.Adapters.Postgres.TypeModule.encode_params/3
         (postgrex) lib/postgrex/query.ex:45: DBConnection.Query.Postgrex.Query.encode/3
    (db_connection) lib/db_connection.ex:1070: DBConnection.describe_run/5
    (db_connection) lib/db_connection.ex:1141: anonymous fn/4 in DBConnection.run_meter/5
    (db_connection) lib/db_connection.ex:1198: DBConnection.run_begin/3
    (db_connection) lib/db_connection.ex:584: DBConnection.prepare_execute/4
             (ecto) lib/ecto/adapters/postgres/connection.ex:80: Ecto.Adapters.Postgres.Connection.prepare_execute/5
             (ecto) lib/ecto/adapters/sql.ex:243: Ecto.Adapters.SQL.sql_call/6
             (ecto) lib/ecto/adapters/sql.ex:431: Ecto.Adapters.SQL.execute_and_cache/7
             (ecto) lib/ecto/repo/queryable.ex:130: Ecto.Repo.Queryable.execute/5
             (ecto) lib/ecto/repo/queryable.ex:35: Ecto.Repo.Queryable.all/4
             (ecto) lib/ecto/repo/queryable.ex:68: Ecto.Repo.Queryable.one/4
```

the migration to create the table is
```elixir
defmodule Xanatos.Repo.Migrations.CreateIpAddress do
  use Ecto.Migration

  def change do
    create table(:ip_addresses, primary_key: false) do
      add :id, :binary_id, primary_key: true
      add :ip, :inet
      add :mac, :macaddr
      add :subnet, :cidr
      timestamps type: :utc_datetime
    end
  end
end
```

my model schema is
```elixir
  schema "ip_addresses" do
    field :ip, EctoNetwork.INET
    field :mac, EctoNetwork.MACADDR
    field :subnet, EctoNetwork.CIDR
    timestamps()
  end
```
I'm not sure if this is a problem here or with ecto_network.

**Comment by josevalim** [comment]
Can you please try this:

```elixir
Repo.query!("SELECT i0.id FROM "ip_addresses" AS i0 WHERE (i0.subnet = $1)", [%Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}])
```

and then this:

```
Repo.query!("SELECT i0.id FROM "ip_addresses" AS i0 WHERE (i0.subnet = $1::cidr)", [%Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}])
```

And let us know what works?

**Comment by darkbushido** [comment]
The quotes around ip_addresses breaks the string, I removed them and for the first query I get
```
** (ArgumentError) Postgrex expected %Postgrex.INET{}, got %Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}. Please make sure the value you are passing matches the definition in your table or in your query or convert the value accordingly.
[debug] QUERY ERROR db=1.4ms
SELECT i0.id FROM ip_addresses AS i0 WHERE (i0.subnet = $1) [%Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}]
             (ecto) /Users/lance/rapid7/Xanatos-poc/xanatos/deps/postgrex/lib/postgrex/type_module.ex:716: Ecto.Adapters.Postgres.TypeModule.encode_params/3
         (postgrex) lib/postgrex/query.ex:45: DBConnection.Query.Postgrex.Query.encode/3
    (db_connection) lib/db_connection.ex:1070: DBConnection.describe_run/5
    (db_connection) lib/db_connection.ex:1141: anonymous fn/4 in DBConnection.run_meter/5
    (db_connection) lib/db_connection.ex:1198: DBConnection.run_begin/3
    (db_connection) lib/db_connection.ex:584: DBConnection.prepare_execute/4
             (ecto) lib/ecto/adapters/postgres/connection.ex:93: Ecto.Adapters.Postgres.Connection.execute/4
             (ecto) lib/ecto/adapters/sql.ex:243: Ecto.Adapters.SQL.sql_call/6
             (ecto) lib/ecto/adapters/sql.ex:193: Ecto.Adapters.SQL.query!/5
```
The second query has the same quoting issue, when escaped or removed it looks like it works.
```
Repo.query!("SELECT i0.id FROM ip_addresses AS i0 WHERE (i0.subnet = $1::cidr)", [%Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}])  
[debug] QUERY OK db=2.3ms
SELECT i0.id FROM ip_addresses AS i0 WHERE (i0.subnet = $1::cidr) [%Postgrex.CIDR{address: {192, 168, 0, 0}, netmask: 24}]
%Postgrex.Result{columns: ["id"], command: :select, connection_id: 62552,
 num_rows: 1,
 rows: [[<<61, 24, 199, 186, 246, 204, 72, 238, 153, 88, 159, 226, 196, 56, 112,
     184>>]]}
```

**Comment by josevalim** [comment]
@ericmj is there something we can do in the extension system here? Should we automatically encode CIDR as INET if Postgres ask for it? They seem to be compatible.

**Comment by ericmj** [comment]
I guess we could but I would like to understand why Postgres does this.

**Comment by josevalim** [comment]
That's a good question. According to the docs, cidr are a subset of inet, so I would suppose it generalizes to the supertype (but still stores them correctly as cidr on read). We could also solve this by potentially having a single data structure.

**Comment by fishcakez** [comment]
We have at least the issue that netmasks should be supported by Postgrex.INET: https://www.postgresql.org/docs/9.6/static/datatype-net-types.html#DATATYPE-INET-VS-CIDR. However to quote the linked section "Any cidr value can be cast to inet __implicitly__ or explicitly". I think we can only handle the implicit cast cleanly by using a single datatype as postgrex won't know when it is an implicit cast.

**Comment by josevalim** [comment]
:+1: for a single data type then.

**Comment by josevalim** [comment]
Fixed in a9486265dcf876b3e3e9e9fcbee101618c7ea01c.
